### PR TITLE
NEO-51318: fix SimpleJson serialization issues

### PR DESCRIPTION
**Pull request by mkiki**
## Description

I discovered a few more issues in the XML to JSON serialization.
More documentation: https://opensource.adobe.com/acc-js-sdk/simpleJson.html
 

when XML element has both a text content and attributes or child elements
when root element has only text
when XML element has both an attribute and a child element with the same name
when array contains element which only have text
when XML is formatted (and therefore has insignificant whitespaces)
XML collections (i.e. when root node ends with “-collection”) would sometimes contain a JSON property “#text”
 

So far the SimpleJSon rules where:

 

XML Attributes are transformed to JSON properties without the “@” prefix. The rationale is that Campaign generally uses attributes, so we make the attribute syntax the most straightforward.
XML Elements are transformed to JSON objects or arrays if the element appears more than once
XML Elements containing only text are transformed to JSON properties with the “$” prefix. The rationale is that Campaign usually does not have XML elements containing both text and children elements and therefore the JSON syntax remains simple in the common case
 

I’m adding the following rules

 

When XML element has both a XML child element and an attribute with the same name, the attribute will have the “@” prefix. This assume that both attribute and element are actually present (possibly empty) in the XML
XML elements containing only text will continue to be transformed to JSON properties with the “$” prefix. Whitespaces in the text content are unchanged.
XML elements containing text and either attributes or children will be transformed to JSON objects having the “$” property which will contain the text. The whitespaces in the text will be removed
When XML element contain multiple XML text nodes, they will be concatenated and seen as one in JSON. This is not recursive. Each text node is considered independently regarding whitespaces
CDATA nodes will be considered as text nodes but for which whitespaces are never removed
Whitespace removal consist on removing the “ “, “\t”, “\n” and “\r” characters at the beginning or the end of the text
 

Here are a few examples of what's expected

 
```
expect(toJSON('<root>Hello</root>')).toEqual({ $: "Hello" });
expect(toJSON('<ctx><delivery>Hello</delivery></ctx>')).toEqual({ "$delivery": "Hello" });
expect(toJSON('<delivery transaction="0">0</delivery>')).toEqual({ "transaction": "0", "$": "0" });
expect(toJSON('<ctx><delivery transaction="0">0</delivery></ctx>')).toEqual({delivery: { "transaction":"0", "$": "0" }});
expect(toJSON('<ctx><delivery>Hello<child name="world"/></delivery></ctx>')).toEqual({delivery: { "$":"Hello", child:
{ name: "world" }
}});
expect(toJSON('<ctx><delivery><child name="world"/>Hello</delivery></ctx>')).toEqual({delivery: { "$":"Hello", child:
{ name: "world" }
}});
expect(toJSON('<root><delivery>Hello<child/>World</delivery></root>')).toEqual({ delivery: { $:"HelloWorld", child:{} } });
```

## Motivation and Context

* Processing report data (see https://github.com/adobe/acc-js-sdk/pull/50)
* The report data payload contain XML element which have both attributes and text content which was not supported before

## How Has This Been Tested?

* New unit tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

This change is actually a breaking change since it changes how JSON serialization works. However, it only changes for cases which were missing / buggy / not properly defined. So it should be backwards compatible. If not, the code that it breaks was probably not working / not safe anyways.

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: docs/simpleJson.html [hunk 1]
```diff
@@ -86,3 +95,66 @@
...
+<p>
+  Whitespace trimming consist of removing all the leading and trainling whitespaces of each text node and concatenating all text node values. If the resulting value is empty, the text node is ignored. The rationale is to remove insignificant spaces created when formatting XML documents
+</p>
+
+
+<h1>Exceptions</h1>
+
+<h2>When an attribute has the same name of an element</h2>
+
+<p>If an element contains another element and an attribute which have the same name, the attribute name is prefixed with the "@" character</p>
```
Comment (rahuljain936) [inline docs/simpleJson.html:123]: Can add a test case for this case as well.

File: src/util.js [hunk 2]
```diff
@@ -133,6 +133,25 @@ class Util {
     }
     return obj;
   }
+
+  /**
+   * Trims the text value of an XML text node, i.e. remove leading and trailing whitespaces (space, tab, \r and \n)
+   * @param {string} text value of the XML text node
+   * @returns {string} the trimmed value
+   */
+  static removeXmlWhiteSpaces(text) {
```
Comment (mukulgupta21) [inline src/util.js:142]: Couldn't we have used `text.trim()` instead?
It handles both whitespace and line terminators.
[Reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/trim#return_value)

File: src/util.js [hunk 3]
```diff
@@ -133,6 +133,25 @@ class Util {
...
+      var first = 0, last = text.length - 1;
+      while (first <= last) {
+        var c = text[first];
+        if (c === ' ' || c === '\t' || c === '\n' || c === '\r') { first = first + 1; continue; }
+        c = text[last];
+        if (c === ' ' || c === '\t' || c === '\n' || c === '\r') { last = last - 1; continue; }
+        break;
+    }
+    text = text.substring(first, last + 1);
+    return text;
```
Comment (mukulgupta21) [inline src/util.js:153]:
```suggestion
    if (!text) return text;
    var first = 0, last = text.length - 1;
    while (first <= last) {
      var c = text[first];
      if (c === ' ' || c === '\t' || c === '\n' || c === '\r') { first = first + 1; continue; }
      c = text[last];
      if (c === ' ' || c === '\t' || c === '\n' || c === '\r') { last = last - 1; continue; }
      break;
    }
    text = text.substring(first, last + 1);
    return text;
```

File: test/domUtil.test.js [hunk 4]
```diff
@@ -177,6 +177,24 @@ describe('DomUtil', function() {
...
+        function fromJSON(json) {
+            var xml = DomUtil.fromJSON("root", json);
+            return DomUtil.toXMLString(xml);
+        }
+
+        it("Should handle fixes made in version 1.1.3", () => {
+            expect(fromJSON({ $: "Hello" })).toBe("<root>Hello</root>");
+            expect(fromJSON({ "$delivery": "Hello" })).toBe("<root><delivery>Hello</delivery></root>");
+            expect(fromJSON({ "delivery": { $: "Hello" } })).toBe("<root><delivery>Hello</delivery></root>");
+            //expect(fromJSON({ "$delivery": "World", "delivery": { $: "Hello" } })).toBe("<root><delivery>Hello</delivery></root>");
```
Comment (mukulgupta21) [inline test/domUtil.test.js:190]: Debug code? Perhaps, your intention is to fix this case in the future. We could leave it like that then.
The asserted behaviour seems correct to me.

File: src/domUtil.js [hunk 5]
```diff
@@ -457,6 +466,46 @@ class DomUtil {
...
         }
+
+        // Proceed with text nodes in SimpleJson format. 
+        if (flavor === "SimpleJson") {
+            var text = "";
+            child = xml.firstChild;
+            while (child) {
+                if (child.nodeType === 3) { // text 
+                    var nodeText = child.nodeValue;
+                    // Whitesapce trimming rule: alwasy trim for the root node, and trim for non-root nodes
```
Comment (mukulgupta21) [inline src/domUtil.js:477]:
```suggestion
                    // Whitespace trimming rule: always trim for the root node, and trim for non-root nodes
```

**Review by mukulgupta21 — CHANGES_REQUESTED** [review]
Can we use [trim](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/trim#return_value) instead of our own implementation.
Other than that, I have only some typos and cosmetic change suggestions.

**Pull request merged**